cmd/juju/caas/add.go

```diff
@@ -195,25 +228,46 @@ func (c *AddCAASCommand) Run(ctxt *cmd.Context) error {
 		return errors.Errorf("CAData attribute should be a string")
 	}
 
-	newCloud := cloud.Cloud{
-		Name:           c.caasName,
-		Type:           c.caasType,
-		Endpoint:       currentCloud.Endpoint,
-		AuthTypes:      []cloud.AuthType{credential.AuthType()},
-		CACertificates: []string{cloudCAData},
+	newCloud := jujucloud.Cloud{
+		Name:            c.caasName,
+		Type:            c.caasType,
+		Endpoint:        currentCloud.Endpoint,
+		AuthTypes:       []jujucloud.AuthType{credential.AuthType()},
+		CACertificates:  []string{cloudCAData},
+		HostCloudRegion: c.cloudRegion,
 	}
 
-	if err := addCloudToLocal(c.cloudMetadataStore, newCloud); err != nil {
-		return errors.Trace(err)
-	}
-
-	cloudClient, err := c.apiFunc()
+	cloudClient, err := c.addCloudAPIFunc()
 	if err != nil {
 		return errors.Trace(err)
 	}
 	defer cloudClient.Close()
 
-	if err := addCloudToController(cloudClient, newCloud); err != nil {
+	if err := c.addCloudToControllerWithRegion(cloudClient, newCloud); err != nil {
+		if params.IsCodeCloudRegionRequired(err) {
+			// try to fetch cloud and region then retry.
+			cloudRegion, err := c.getClusterRegion(ctx, newCloud, credential)
+			errMsg := `
+Jaas requires cloud and region information. But it's
+not possible to fetch cluster region in this case, 
+	please use --region to specify the cloud/region manually
+`[1:]
+			if err != nil {
+				return errors.Annotate(err, errMsg)
+			}
+			if cloudRegion == "" {
+				return errors.NewNotValid(nil, errMsg)
+			}
+			newCloud.HostCloudRegion = cloudRegion
+			if err := c.addCloudToControllerWithRegion(cloudClient, newCloud); err != nil {
+				return errors.Trace(err)
+			}
+		} else {
```

Comment:
We can reduce the amount of indentation by inverting the if condition
```
  if !params.IsCodeCloudRegionRequired(err) {
    return errors.Trace(err)
  }
  // try to fetch cloud and region then retry.
  ....
```

Reply:
yeah, good finding, changed